golang/cosmos/x/swingset/keeper/store_snapshotter.go

```diff
@@ -0,0 +1,601 @@
+// these operations, and which is provided to InitiateSnapshot.
+type SwingStoreSnapshotTaker interface {
+	// Called after the SwingStore snapshot was initiated in a goroutine
+	// Must call the `retrieveSwingStoreExport` function before returning, which
+	// will in turn synchronously invoke `ProcessExportedSnapshot` once the
+	// snapshot is ready
+	TakeSnapshot(height uint64, retrieveSwingStoreExport func() error) error
+	// Called when the SwingStore snapshot has been retrieved
+	// The implementation must synchronously consume the provider, which becomes
+	// invalid after the method returns.
+	ProcessExportedSnapshot(provider SwingStoreExportProvider) error
+}
+
+type snapshotAction struct {
+	Type        string            `json:"type"` // COSMOS_SNAPSHOT
+	BlockHeight uint64            `json:"blockHeight"`
+	Request     string            `json:"request"` // "initiate", "discard", "retrieve", or "restore"
+	Args        []json.RawMessage `json:"args,omitempty"`
+}
+
+// SwingStoreExportOptions are configurable options provided to the JS SwingStore export
+type SwingStoreExportOptions struct {
+	// The export mode can be "current", "archival" or "debug"
+	// See packages/cosmic-swingset/src/export-kernel-db.js `initiateSwingStoreExport` and
+	// packages/swing-store/src/swingStore.js `makeSwingStoreExporter`
+	ExportMode string `json:"exportMode,omitempty"`
+	// A flag indicating whether "export data" should be part of the SwingStore export
+	// If false, the resulting SwingStoreExportProvider's GetExportData will
+	// error with io.EOF
+	IncludeExportData bool `json:"includeExportData,omitempty"`
+}
+
+// SwingStoreImportOptions are configurable options provided to the JS SwingStore import
+type SwingStoreImportOptions struct {
+	// A flag indicating whether the SwingStore import should attempt to load
+	// all historical artifacts available from the export provider
+	IncludeHistorical bool `json:"includeHistorical,omitempty"`
+}
+
+type restoreOptions struct {
+	ExportDir         string `json:"exportDir"`
+	IncludeHistorical bool   `json:"includeHistorical,omitempty"`
+}
+
+// SwingStoreSnapshotter manages the exclusive communication with the JS side
+// for SwingStore snapshots, ensuring insensitivity to sub-block timing, and
+// enforcing concurrency requirements.
+// The caller of this submodule must arrange block level commit synchronization,
+// to ensure the results are deterministic.
+//
+// Some `blockingSend` calls performed by this submodule are non-deterministic.
+// This submodule will send messages to JS from goroutines at unpredictable
+// times, but this is safe because when handling the messages, the JS side
+// does not perform operations affecting consensus and ignores state changes
+// since committing the previous block.
+// Some other `blockingSend` calls however do change the JS swing-store and
+// must happen before the Swingset controller on the JS side was inited.
```

Comment:
How are the `blockingSend` calls that must happen before JS Swingset controller init distinguished here?

Reply:
The `blockingSend` function takes a `mustNotBeInited` param?